jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+                    LOG.debug("not upgraded no key {}",baseRequest);
+                return false;
+            }
+
+            // Create a connection from a connection factory (from context or connector)
+            ServletContext context = baseRequest.getServletContext();
+            WebSocketConnectionFactory connectionFactory = null;
+            if (context!=null)
+                connectionFactory = (WebSocketConnectionFactory)context.getAttribute(WebSocketConnectionFactory.class.getName());
+            if (connectionFactory==null)
+                connectionFactory = baseRequest.getHttpChannel().getConnector().getBean(WebSocketConnectionFactory.class);
+            if (connectionFactory==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection factory {}",baseRequest);
+                return false;
+            }
+            HttpChannel channel = baseRequest.getHttpChannel();
+            WebSocketCoreConnection connection =
+                    connectionFactory.newConnection(channel.getConnector(),channel.getEndPoint(),extensionCSVs.getValues());
+            if (connection==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection {}",baseRequest);
+            }
+
+            // Create a session from a session factory (from context or connector)
+            WebSocketSessionFactory sessionFactory = null;
+            if (context!=null)
+                sessionFactory = (WebSocketSessionFactory)context.getAttribute(WebSocketSessionFactory.class.getName());
+            if (sessionFactory==null)
+                sessionFactory = baseRequest.getHttpChannel().getConnector().getBean(WebSocketSessionFactory.class);
+            if (sessionFactory==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no session factory {}",baseRequest);
+                return false;
+            }
+            WebSocketCoreSession session = sessionFactory.newSession(connection, subprotocol);
+            if (session==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no session {}",baseRequest);
+            }
+
+            connection.setSession(session);
+
+            // send upgrade response
+            Response baseResponse = baseRequest.getResponse();
+            baseResponse.setStatus(HttpServletResponse.SC_SWITCHING_PROTOCOLS);
+            baseResponse.getHttpFields().add(UpgradeWebSocket);
+            baseResponse.getHttpFields().add(ConnectionUpgrade);
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_ACCEPT, AcceptHash.hashKey(key));
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_EXTENSIONS,
+                                             ExtensionConfig.toHeaderValue(connection.getExtensionStack().getNegotiatedExtensions()));
```

Comment:
If you need the connection, then you are doing this in the wrong order.
The upgrade occurs before the websocket specific connection is even created.

Reply:
@joakime the upgrade cannot occur before the WebSocket connection is created, as we should call `endPoint.upgrade(Connection)` and pass it the new connection.

Reply:
ExtensionStack is the post-negotiated reality of the active connection.  
Using it to perform the negotiation is incorrect.

Reply:
I don't think the implementation matters either way.  We can send upgrade response and then do the upgrade or the other way around, because we are not fill interested during this period.

I did it this way just so that if there is any error in generating the upgrade response then we do not do the upgrade.

See comments above regarding extension stack